walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,179 @@
+    /* Loop through all the events our optional attendees are attending and
+       find free time slots we can have for our meeting */
+    for (Event event : optionalAttendeeEvents) {
+      findConflictAndResolve(optionalFreeTime, event);
+    }
+    /* If we can't find any free time slots with optional attendees
+      then return the original freeTime slots */
+    if (optionalFreeTime.isEmpty()) return freeTime;
+
+    return optionalFreeTime;
+  }
+
+  // This method finds the overlapping free time slot and event; altering the freeTime so that it
+  // doesn't overlap anymore.
+  private void findConflictAndResolve(
+      LinkedList<TimeRange> time, Event event) {
+      // Go through all the free time slots and check from which one is the event overlapping
+      ListIterator<TimeRange> iterator = time.listIterator();
+      while (iterator.hasNext()) {
+        TimeRange freeSlot = iterator.next();
+        // When th eoverlapping free slot is found send our free time and event slot to get altered.
+        if (freeSlot.overlaps(event.getWhen())) {
+          updateFreeTime(freeSlot, event.getWhen(), time, iterator);
+          // After updating free time is done, check if the event ends in this free slots
+          if (event.getWhen().end() <= freeSlot.end()) break;
+        }
+    }
+  }
+
+  /* This method free time, free time slot and current event with our free time slot and the iterator
+    and checks all conditions in which the event and time slot could overlap. 
+    Then calculates the free time minus overlapping part. 
+    It checks the new slot with the meeting slot and then adds or removes it. 
+    (By removing/ setting or adding in the iterator). */
+  private void updateFreeTime(
+      TimeRange freeSlot,
+      TimeRange eventSlot,
+      LinkedList<TimeRange> freeTime,
+      ListIterator<TimeRange> iterator) {
+
+    int freeSlotStart = freeSlot.start();
+    int freeSlotEnd = freeSlot.end();
+
+    int eventSlotStart = eventSlot.start();
+    int eventSlotEnd = eventSlot.end();
+
+    // Case: Free time: |---|
+    // Event slot:    |-------|
+    if (eventSlotStart <= freeSlotStart && eventSlotEnd >= freeSlotEnd) {
+      iterator.remove();
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:     |-----|
+    // or
+    // |------|
+    // |--|
+    else if (eventSlotStart <= freeSlotStart && eventSlotEnd < freeSlotEnd) {
+      TimeRange newSlot = TimeRange.fromStartEnd(eventSlotEnd, freeSlotEnd, false);
+      if (isSlotTimeEnough(newSlot)) iterator.set(newSlot);
+      else iterator.remove();
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:            |-----|
+    // or
+    // |------|
+    //     |--|
+    else if (eventSlotStart > freeSlotStart && eventSlotEnd >= freeSlotEnd) {
+      TimeRange newSlot = TimeRange.fromStartEnd(freeSlotStart, eventSlotStart, false);
+      if (isSlotTimeEnough(newSlot)) iterator.set(newSlot);
+      else iterator.remove();
+
+    }
+
```

Comment:
Moving all the comments within "else if" so that those code blocks are connected.

Reply:
Done!